File: formlogic/backend/src/Services/EnvelopeValidator.php
<?php

declare(strict_types=1);

namespace FormLogic\Services;

use Seld\JsonLint\JsonParser;
use Seld\JsonLint\ParsingException;

/**
 * Structural validation of __flenc:1 response envelopes (docs/E2EE_PRIVATE_FORMS_PLAN.md §6).
 *
 * NEVER decrypts anything — the server holds no content keys. Duplicate JSON keys are
 * invisible after json_decode (it keeps the last), so the RAW request body is parsed with
 * seld/jsonlint's DETECT_KEY_CONFLICTS, which walks every nesting level (request root AND
 * envelope object). Fail closed: any deviation rejects the request before storage.
 */
final class EnvelopeValidator
{
    public const ENVELOPE_VERSION = 1;
    public const CONTENT_SUITE = 'xchacha20p1305.1';
    public const WRAP_SUITE = 'sealedbox-x25519xsalsa20p1305.1';

    /** Serialized-envelope storage cap = existing MAX_ANSWER_BYTES. */
    public const MAX_ENVELOPE_BYTES = 2_000_000;
    /** Raw request cap: envelope + idempotency key + attachment claims headroom. */
    public const MAX_REQUEST_BYTES = 2_100_000;
    public const MAX_CT_B64_CHARS = 1_900_000;
    public const WRAPPED_DEK_BYTES = 80;
    public const NONCE_BYTES = 24;
    public const MAX_ENVELOPE_KEYS = 24;
    public const MAX_ATTACHMENTS = 50;

    private const ALLOWED_ENVELOPE_KEYS = [
        '__flenc', 'recordId', 'rev', 'keyId', 'epoch', 'content', 'wrap',
        'schemaVersion', 'schemaHash', 'attachments', 'wrappedDek', 'nonce', 'ct',
    ];

    private const UUID_V4_RE = '/^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/';
    private const KEY_ID_RE = '/^fik_[a-zA-Z0-9_-]{1,36}$/';
    private const FILE_ID_RE = '/^fil_[a-zA-Z0-9_-]{1,36}$/';
    private const SCHEMA_HASH_RE = '/^[0-9a-f]{64}$/';

    /**
     * Parses a raw private-form request body with duplicate-key detection at every level.
     *
     * @return array{0: array<string,mixed>|null, 1: string|null, 2: string|null}
     *         [decodedBody, errorCode, errorMessage]
     */
    public function parseRequestBody(string $raw): array
    {
        if (strlen($raw) > self::MAX_REQUEST_BYTES) {
            return [null, 'payload_too_large', 'request body exceeds the private-form size cap'];
        }
        $parser = new JsonParser();
        try {
            $decoded = $parser->parse($raw, JsonParser::DETECT_KEY_CONFLICTS | JsonParser::PARSE_TO_ASSOC);
        } catch (ParsingException $e) {
            $isDup = str_contains(strtolower($e->getMessage()), 'duplicate');
            return [null, 'envelope_invalid', $isDup ? 'duplicate JSON keys are not allowed' : 'request body is not valid JSON'];
        }
        if (!is_array($decoded) || array_is_list($decoded)) {
            return [null, 'envelope_invalid', 'request body must be a JSON object'];
        }
        return [$decoded, null, null];
    }

    /**
     * Structural envelope validation (no decryption).
     *
     * @param mixed $env decoded envelope value from the request body
     * @param string $formId
     * @param list<array{key_id:string, ingest_epoch:int, schema_version:int, schema_hash:string}> $acceptableManifests
     *        Manifest tuples currently acceptable for this form (plan §8: active, or
     *        retiring with now < accept_until — resolved by the caller).
     * @param int|null $expectedRev null = create (rev must be 1); else rev must equal expectedRev + 1
     * @return array{ok: bool, code?: string, message?: string, envelope?: array<string,mixed>}
     */
    public function validateEnvelope(mixed $env, string $formId, array $acceptableManifests, ?int $expectedRev = null): array
    {
        if (!is_array($env) || array_is_list($env)) {
            return $this->fail('envelope_invalid', 'envelope must be a JSON object');
        }
        if (count($env) > self::MAX_ENVELOPE_KEYS) {
            return $this->fail('envelope_invalid', 'envelope has too many keys');
        }
        foreach (array_keys($env) as $key) {
            if (!in_array($key, self::ALLOWED_ENVELOPE_KEYS, true)) {
                return $this->fail('envelope_invalid', "unexpected envelope key: {$key}");
            }
        }
        if (($env['__flenc'] ?? null) !== self::ENVELOPE_VERSION) {
            return $this->fail('envelope_invalid', 'unsupported envelope version');
        }
        if (($env['content'] ?? null) !== self::CONTENT_SUITE) {
            return $this->fail('envelope_invalid', 'unknown content suite');
        }
        if (($env['wrap'] ?? null) !== self::WRAP_SUITE) {
            return $this->fail('envelope_invalid', 'unknown wrap suite');
        }
        $recordId = $env['recordId'] ?? null;
        if (!is_string($recordId) || preg_match(self::UUID_V4_RE, $recordId) !== 1) {
            return $this->fail('envelope_invalid', 'recordId must be a UUIDv4');
        }
        $rev = $env['rev'] ?? null;
        if (!is_int($rev) || $rev < 1) {
            return $this->fail('envelope_invalid', 'rev must be a positive integer');
        }
        if ($expectedRev === null) {
            if ($rev !== 1) {
                return $this->fail('envelope_invalid', 'new responses must carry rev 1');
            }
        } elseif ($rev !== $expectedRev + 1) {
            return $this->fail('revision_conflict', 'envelope rev must be expectedRev + 1');
        }
        $keyId = $env['keyId'] ?? null;
        if (!is_string($keyId) || preg_match(self::KEY_ID_RE, $keyId) !== 1) {
            return $this->fail('envelope_invalid', 'keyId malformed');
        }
        $epoch = $env['epoch'] ?? null;
        if (!is_int($epoch) || $epoch < 1) {
            return $this->fail('envelope_invalid', 'epoch must be a positive integer');
        }
        $schemaVersion = $env['schemaVersion'] ?? null;
        if (!is_int($schemaVersion) || $schemaVersion < 1) {
            return $this->fail('envelope_invalid', 'schemaVersion must be a positive integer');
        }
        $schemaHash = $env['schemaHash'] ?? null;
        if (!is_string($schemaHash) || preg_match(self::SCHEMA_HASH_RE, $schemaHash) !== 1) {
            return $this->fail('envelope_invalid', 'schemaHash must be 64 lowercase hex chars');
        }

        if (array_key_exists('attachments', $env)) {
            $atts = $env['attachments'];
            if (!is_array($atts) || !array_is_list($atts) || $atts === []) {
                return $this->fail('envelope_invalid', 'attachments must be a non-empty list when present');
            }
            if (count($atts) > self::MAX_ATTACHMENTS) {
                return $this->fail('envelope_invalid', 'too many attachments');
            }
            $seen = [];
            $prev = null;
            foreach ($atts as $id) {
                if (!is_string($id) || preg_match(self::FILE_ID_RE, $id) !== 1) {
                    return $this->fail('envelope_invalid', 'attachment id malformed');
                }
                if (isset($seen[$id])) {
                    return $this->fail('envelope_invalid', 'duplicate attachment id');
                }
                if ($prev !== null && strcmp($prev, $id) > 0) {
                    return $this->fail('envelope_invalid', 'attachments must be sorted');
                }
                $seen[$id] = true;
                $prev = $id;
            }
        }

        $wrappedDek = $this->decodeExactB64($env['wrappedDek'] ?? null, self::WRAPPED_DEK_BYTES);
        if ($wrappedDek === null) {
            return $this->fail('envelope_invalid', 'wrappedDek must be canonical base64 of exactly 80 bytes');
        }
        $nonce = $this->decodeExactB64($env['nonce'] ?? null, self::NONCE_BYTES);
        if ($nonce === null) {
            return $this->fail('envelope_invalid', 'nonce must be canonical base64 of exactly 24 bytes');
        }
        $ct = $env['ct'] ?? null;
        if (!is_string($ct) || $ct === '' || strlen($ct) > self::MAX_CT_B64_CHARS || !$this->isCanonicalB64($ct)) {
            return $this->fail('envelope_invalid', 'ct must be canonical base64 within the size cap');
        }

        $serialized = json_encode($env, JSON_UNESCAPED_SLASHES);
        if ($serialized === false || strlen($serialized) > self::MAX_ENVELOPE_BYTES) {
            return $this->fail('payload_too_large', 'sealed envelope exceeds the 2 MB storage cap');
        }

        $tupleOk = false;
        foreach ($acceptableManifests as $m) {
            if ($m['key_id'] === $keyId
                && $m['ingest_epoch'] === $epoch
                && $m['schema_version'] === $schemaVersion
                && hash_equals($m['schema_hash'], $schemaHash)
            ) {
                $tupleOk = true;
                break;
            }
        }
        if (!$tupleOk) {
            return $this->fail('key_epoch_retired', 'envelope does not match any currently acceptable manifest for this form');
        }

        // formId is not an envelope field (it is bound inside the client AAD); the caller
        // resolves manifests strictly for $formId, so a cross-form envelope fails the tuple
        // check above. $formId stays a parameter to make that contract explicit.
        unset($formId);

        return ['ok' => true, 'envelope' => $env];
    }

    /** Canonical padded base64 check: decode strictly AND re-encode must round-trip. */
    private function isCanonicalB64(string $s): bool
    {
        if ($s === '' || strlen($s) % 4 !== 0) {
            return false;
        }
        $decoded = base64_decode($s, true);
        if ($decoded === false) {
            return false;
        }
        return base64_encode($decoded) === $s;
    }

    private function decodeExactB64(mixed $value, int $expectedBytes): ?string
    {
        if (!is_string($value) || !$this->isCanonicalB64($value)) {
            return null;
        }
        $decoded = base64_decode($value, true);
        if ($decoded === false || strlen($decoded) !== $expectedBytes) {
            return null;
        }
        return $decoded;
    }

    /** @return array{ok: bool, code: string, message: string} */
    private function fail(string $code, string $message): array
    {
        return ['ok' => false, 'code' => $code, 'message' => $message];
    }
}

File: formlogic/desktop/src-tauri/src/data/envelope_validator.rs
//! Independent Desktop validation of `__flenc:1` response envelopes
//! (plan §12.2). The Desktop primary never trusts the Cloud to have validated
//! a Private response: this mirrors backend/src/Services/EnvelopeValidator.php
//! field-for-field (allowlists, suites, canonical base64, size caps, manifest
//! tuple matching) on top of the duplicate-key-rejecting
//! [`super::strict_json`] parser. Shared adversarial fixtures:
//! docs/contracts/data-envelope-adversarial.json (asserted here AND by
//! backend/tests/Unit/DataEnvelopeAdversarialTest.php).
//!
//! NEVER decrypts anything — a storage node holds no content keys (plan D3).

use serde_json::Value;

pub const ENVELOPE_VERSION: i64 = 1;
pub const CONTENT_SUITE: &str = "xchacha20p1305.1";
pub const WRAP_SUITE: &str = "sealedbox-x25519xsalsa20p1305.1";

pub const MAX_ENVELOPE_BYTES: usize = 2_000_000;
pub const MAX_REQUEST_BYTES: usize = 2_100_000;
pub const MAX_CT_B64_CHARS: usize = 1_900_000;
pub const WRAPPED_DEK_BYTES: usize = 80;
pub const NONCE_BYTES: usize = 24;
pub const MAX_ENVELOPE_KEYS: usize = 24;
pub const MAX_ATTACHMENTS: usize = 50;

const ALLOWED_ENVELOPE_KEYS: &[&str] = &[
    "__flenc", "recordId", "rev", "keyId", "epoch", "content", "wrap",
    "schemaVersion", "schemaHash", "attachments", "wrappedDek", "nonce", "ct",
];

/// One acceptable manifest tuple for the form (active, or retiring within its
/// signed grace window — resolved by the caller, plan §12.2).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ManifestTuple {
    pub key_id: String,
    pub ingest_epoch: i64,
    pub schema_version: i64,
    pub schema_hash: String,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct EnvelopeError {
    pub code: &'static str,
    pub message: String,
}

fn fail(code: &'static str, message: impl Into<String>) -> EnvelopeError {
    EnvelopeError { code, message: message.into() }
}

/// Parse a raw private-form request body with duplicate-key detection at every
/// level (mirror of EnvelopeValidator::parseRequestBody).
pub fn parse_request_body(raw: &[u8]) -> Result<Value, EnvelopeError> {
    if raw.len() > MAX_REQUEST_BYTES {
        return Err(fail("payload_too_large", "request body exceeds the private-form size cap"));
    }
    let value = super::strict_json::parse(raw).map_err(|e| {
        if e.contains("duplicate") {
            fail("envelope_invalid", "duplicate JSON keys are not allowed")
        } else {
            fail("envelope_invalid", "request body is not valid JSON")
        }
    })?;
    if !value.is_object() {
        return Err(fail("envelope_invalid", "request body must be a JSON object"));
    }
    Ok(value)
}

/// Structural envelope validation (no decryption) — mirror of
/// EnvelopeValidator::validateEnvelope. `expected_rev = None` means create
/// (rev must be 1); otherwise rev must equal `expected_rev + 1`.
pub fn validate_envelope(
    env: &Value,
    acceptable_manifests: &[ManifestTuple],
    expected_rev: Option<i64>,
) -> Result<(), EnvelopeError> {
    let Some(map) = env.as_object() else {
        return Err(fail("envelope_invalid", "envelope must be a JSON object"));
    };
    if map.len() > MAX_ENVELOPE_KEYS {
        return Err(fail("envelope_invalid", "envelope has too many keys"));
    }
    for key in map.keys() {
        if !ALLOWED_ENVELOPE_KEYS.contains(&key.as_str()) {
            return Err(fail("envelope_invalid", format!("unexpected envelope key: {key}")));
        }
    }
    if map.get("__flenc").and_then(Value::as_i64) != Some(ENVELOPE_VERSION)
        || !map.get("__flenc").map(Value::is_i64).unwrap_or(false)
    {
        return Err(fail("envelope_invalid", "unsupported envelope version"));
    }
    if map.get("content").and_then(Value::as_str) != Some(CONTENT_SUITE) {
        return Err(fail("envelope_invalid", "unknown content suite"));
    }
    if map.get("wrap").and_then(Value::as_str) != Some(WRAP_SUITE) {
        return Err(fail("envelope_invalid", "unknown wrap suite"));
    }
    let record_id = map.get("recordId").and_then(Value::as_str).unwrap_or("");
    if !is_uuid_v4(record_id) {
        return Err(fail("envelope_invalid", "recordId must be a UUIDv4"));
    }
    let rev = match map.get("rev") {
        Some(v) if v.is_i64() => v.as_i64().unwrap_or(0),
        _ => return Err(fail("envelope_invalid", "rev must be a positive integer")),
    };
    if rev < 1 {
        return Err(fail("envelope_invalid", "rev must be a positive integer"));
    }
    match expected_rev {
        None => {
            if rev != 1 {
                return Err(fail("envelope_invalid", "new responses must carry rev 1"));
            }
        }
        Some(expected) => {
            if rev != expected + 1 {
                return Err(fail("revision_conflict", "envelope rev must be expectedRev + 1"));
            }
        }
    }
    let key_id = map.get("keyId").and_then(Value::as_str).unwrap_or("");
    if !is_key_id(key_id) {
        return Err(fail("envelope_invalid", "keyId malformed"));
    }
    let epoch = match map.get("epoch") {
        Some(v) if v.is_i64() => v.as_i64().unwrap_or(0),
        _ => return Err(fail("envelope_invalid", "epoch must be a positive integer")),
    };
    if epoch < 1 {
        return Err(fail("envelope_invalid", "epoch must be a positive integer"));
    }
    let schema_version = match map.get("schemaVersion") {
        Some(v) if v.is_i64() => v.as_i64().unwrap_or(0),
        _ => return Err(fail("envelope_invalid", "schemaVersion must be a positive integer")),
    };
    if schema_version < 1 {
        return Err(fail("envelope_invalid", "schemaVersion must be a positive integer"));
    }
    let schema_hash = map.get("schemaHash").and_then(Value::as_str).unwrap_or("");
    if !is_hex64(schema_hash) {
        return Err(fail("envelope_invalid", "schemaHash must be 64 lowercase hex chars"));
    }

    if let Some(atts_value) = map.get("attachments") {
        let Some(atts) = atts_value.as_array() else {
            return Err(fail("envelope_invalid", "attachments must be a non-empty list when present"));
        };
        if atts.is_empty() {
            return Err(fail("envelope_invalid", "attachments must be a non-empty list when present"));
        }
        if atts.len() > MAX_ATTACHMENTS {
            return Err(fail("envelope_invalid", "too many attachments"));
        }
        let mut prev: Option<&str> = None;
        for id_value in atts {
            let Some(id) = id_value.as_str() else {
                return Err(fail("envelope_invalid", "attachment id malformed"));
            };
            if !is_file_id(id) {
                return Err(fail("envelope_invalid", "attachment id malformed"));
            }
            if let Some(p) = prev {
                if p == id {
                    return Err(fail("envelope_invalid", "duplicate attachment id"));
                }
                if p > id {
                    return Err(fail("envelope_invalid", "attachments must be sorted"));
                }
            }
            prev = Some(id);
        }
    }

    if decode_exact_b64(map.get("wrappedDek"), WRAPPED_DEK_BYTES).is_none() {
        return Err(fail("envelope_invalid", "wrappedDek must be canonical base64 of exactly 80 bytes"));
    }
    if decode_exact_b64(map.get("nonce"), NONCE_BYTES).is_none() {
        return Err(fail("envelope_invalid", "nonce must be canonical base64 of exactly 24 bytes"));
    }
    let ct = map.get("ct").and_then(Value::as_str).unwrap_or("");
    if ct.is_empty() || ct.len() > MAX_CT_B64_CHARS || !is_canonical_b64(ct) {
        return Err(fail("envelope_invalid", "ct must be canonical base64 within the size cap"));
    }

    let serialized = serde_json::to_string(env)
        .map_err(|_| fail("envelope_invalid", "envelope does not serialize"))?;
    if serialized.len() > MAX_ENVELOPE_BYTES {
        return Err(fail("payload_too_large", "sealed envelope exceeds the 2 MB storage cap"));
    }

    let tuple_ok = acceptable_manifests.iter().any(|m| {
        m.key_id == key_id
            && m.ingest_epoch == epoch
            && m.schema_version == schema_version
            && constant_time_eq(m.schema_hash.as_bytes(), schema_hash.as_bytes())
    });
    if !tuple_ok {
        return Err(fail(
            "key_epoch_retired",
            "envelope does not match any currently acceptable manifest for this form",
        ));
    }
    Ok(())
}

fn is_uuid_v4(s: &str) -> bool {
    let bytes = s.as_bytes();
    if bytes.len() != 36 {
        return false;
    }
    for (i, &b) in bytes.iter().enumerate() {
        match i {
            8 | 13 | 18 | 23 => {
                if b != b'-' {
                    return false;
                }
            }
            14 => {
                if b != b'4' {
                    return false;
                }
            }
            19 => {
                if !matches!(b, b'8' | b'9' | b'a' | b'b') {
                    return false;
                }
            }
            _ => {
                if !b.is_ascii_hexdigit() || b.is_ascii_uppercase() {
                    return false;
                }
            }
        }
    }
    true
}

fn is_token_tail(s: &str, max: usize) -> bool {
    !s.is_empty()
        && s.len() <= max
        && s.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'_' || b == b'-')
}

fn is_key_id(s: &str) -> bool {
    s.strip_prefix("fik_").map(|t| is_token_tail(t, 36)).unwrap_or(false)
}

fn is_file_id(s: &str) -> bool {
    s.strip_prefix("fil_").map(|t| is_token_tail(t, 36)).unwrap_or(false)
}

fn is_hex64(s: &str) -> bool {
    s.len() == 64 && s.bytes().all(|b| b.is_ascii_digit() || (b'a'..=b'f').contains(&b))
}

/// Canonical padded base64: strict decode AND re-encode must round-trip
/// (mirror of EnvelopeValidator::isCanonicalB64).
fn is_canonical_b64(s: &str) -> bool {
    use base64::engine::general_purpose::STANDARD as B64;
    use base64::Engine;
    if s.is_empty() || s.len() % 4 != 0 {
        return false;
    }
    match B64.decode(s) {
        Ok(decoded) => B64.encode(&decoded) == s,
        Err(_) => false,
    }
}

fn decode_exact_b64(value: Option<&Value>, expected: usize) -> Option<Vec<u8>> {
    use base64::engine::general_purpose::STANDARD as B64;
    use base64::Engine;
    let s = value?.as_str()?;
    if !is_canonical_b64(s) {
        return None;
    }
    let decoded = B64.decode(s).ok()?;
    if decoded.len() != expected {
        return None;
    }
    Some(decoded)
}

fn constant_time_eq(a: &[u8], b: &[u8]) -> bool {
    if a.len() != b.len() {
        return false;
    }
    let mut diff = 0u8;
    for (x, y) in a.iter().zip(b.iter()) {
        diff |= x ^ y;
    }
    diff == 0
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;
    use std::path::PathBuf;

    fn fixtures() -> Value {
        let path = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
            .join("../../../docs/contracts/data-envelope-adversarial.json");
        let raw = std::fs::read_to_string(&path)
            .unwrap_or_else(|e| panic!("fixture file missing at {}: {e}", path.display()));
        serde_json::from_str(&raw).expect("fixture file parses")
    }

    fn tuples(case: &Value) -> Vec<ManifestTuple> {
        case["manifests"]
            .as_array()
            .expect("manifests")
            .iter()
            .map(|m| ManifestTuple {
                key_id: m["key_id"].as_str().unwrap().to_string(),
                ingest_epoch: m["ingest_epoch"].as_i64().unwrap(),
                schema_version: m["schema_version"].as_i64().unwrap(),
                schema_hash: m["schema_hash"].as_str().unwrap().to_string(),
            })
            .collect()
    }

    /// Shared adversarial corpus — Rust and PHP must agree case-for-case
    /// (plan §12.2: "Share adversarial fixture corpora across TypeScript,
    /// PHP, and Rust").
    #[test]
    fn adversarial_fixtures_agree_with_php() {
        let f = fixtures();
        let cases = f["cases"].as_array().expect("cases");
        assert!(cases.len() >= 10, "corpus should stay meaningful");
        for case in cases {
            let name = case["name"].as_str().unwrap();
            let raw = case["body"].as_str().unwrap().as_bytes();
            let expected_rev = case["expectedRev"].as_i64();
            let expect_ok = case["ok"].as_bool().unwrap();
            let result = parse_request_body(raw).and_then(|body| {
                let env = body
                    .get("envelope")
                    .cloned()
                    .unwrap_or(Value::Null);
                validate_envelope(&env, &tuples(case), expected_rev)
            });
            match (expect_ok, result) {
                (true, Ok(())) => {}
                (false, Err(e)) => {
                    if let Some(code) = case["code"].as_str() {
                        assert_eq!(e.code, code, "fixture {name}: wrong error code");
                    }
                }
                (true, Err(e)) => panic!("fixture {name}: expected ok, got {}: {}", e.code, e.message),
                (false, Ok(())) => panic!("fixture {name}: expected rejection, got ok"),
            }
        }
    }

    #[test]
    fn helper_predicates() {
        assert!(is_uuid_v4("7d444840-9dc0-41a2-8da8-ff8cb9fca735"));
        assert!(!is_uuid_v4("7D444840-9DC0-41A2-8DA8-FF8CB9FCA735"), "uppercase rejected");
        assert!(!is_uuid_v4("7d444840-9dc0-11a2-8da8-ff8cb9fca735"), "not v4");
        assert!(is_key_id("fik_sample01"));
        assert!(!is_key_id("fik_"));
        assert!(is_canonical_b64("QUJD"));
        assert!(!is_canonical_b64("QUJD "), "whitespace");
        assert!(!is_canonical_b64("QUJDRA=A"), "misplaced padding");
        assert!(!is_canonical_b64("QQ=A"), "non-canonical");
    }
}
